index.ts

```diff
@@ -1,35 +1,65 @@
 export class OatyArray {
-  private _original: object[]
-  private _keys: string[]
-  private _transposed: object = {}
+  private keys: string[] | undefined
+  private original: object[] = []
+  private transposed: { [key: string]: any } = {}
+
+  constructor(options?: { data?: object[], keys?: string[] }) {
+    if (options !== undefined) {
+      if (options.data !== undefined) { this.push(...options.data) }
+      if (options.keys !== undefined) { this.keys = options.keys }
+    }
+  }
 
   get length(): number {
-    return this._original.length
+    return this.original.length
   }
 
-  constructor(input: object[], keys: string[]) {
-    this._original = input.map((x) => Object.assign({}, x))
-    this._keys = keys
+  public getKeys(): string[] | undefined {
+    return this.keys
+  }
 
-    this._transposed = transpose(this._keys, this._original)
+  public getOriginal(): object[] {
+    return this.original
   }
 
-  public get(keyName: string, keyValue: string): object[] {
-    return this._transposed[keyName][keyValue]
+  public getTransposed(): object[] {
+    return this.original
   }
 
-  public push(...items: object[]): number {
-    this._transposed = transpose(this._keys, items, this._transposed)
-    return this._original.push(...items)
+  public get(keyName: string, keyValue: string): object[] | undefined {
+    if (this.transposed === undefined
+     || this.transposed[keyName] === undefined) {
+      return undefined
+    }
+    return this.transposed[keyName][keyValue]
   }
-}
 
-const transpose = (keys: string[], items: object[], current?: object) => {
-  return keys.reduce((transposed, key) => {
-    transposed[key] = items.reduce((acc, item) => {
-      acc[item[key]] ? acc[item[key]].push(item) : acc[item[key]] = [item]
-      return acc
-    }, current || {})
-    return transposed
-  }, {})
+  public push(...data: object[]) {
+    data.forEach((datum) => {
+      const clone = Object.assign({}, datum)
+      this.original.push(clone)
+      this.transpose(clone)
+    })
+    return this.length
+  }
+
+  private getKeysForObject(datum: object): string[] {
+    return this.keys !== undefined ? this.keys : Object.keys(datum)
+  }
+
+  private transpose(datum: { [key: string]: any }) {
+    this.getKeysForObject(datum).forEach((key) => {

```

Comment:
why does the `transpose` use a `forEach` rather that a `reduce` now? I feel like this might have a performance impact

Reply:
Read somewhere that reduce, filter, map, and forEach all have roughly the same performance but that the others transform the data, imo using reduce was somewhat misleading.

Reply:
the `reduce` was transforming the array down to a single value: the transformed object - this is exactly the kind of thing reduce is for. would rather this change is reverted

Reply:
Right but you had to use two `reduces` to get the same effect as one `forEach`

Reply:
Continued in next review